docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+
+This storage plugin will respond "yes" when asked if it owns a volume spec that contains a PV + `sticky-host` annotation, and "no" in every other case. (TODO: do we need to bind mount the directory for every pod?).
+
+## Example
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: test
+  annotations:
+    # Set by scheduler or admin
+    volume.alpha.kubernetes.io/sticky-host: ""
+spec:
+  capacity:
+    storage: 10Mi
+  accessModes:
+    - ReadWriteMany
+  emptyDir: {}
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test
+  annotations:
+    # Mentioned for the sake of clarity
+    volume.alpha.kubernetes.io/storage-class: free
+spec:
+  accessModes: [ "ReadWriteMany" ]
+  resources:
+    requests:
+      storage: 10Mi
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: test-
+spec:
+  terminationGracePeriodSeconds: 0
+  containers:
+  - name: busybox
+    image: busybox
+    command:
+    - sh
+    - -c
+    - touch /data/`hostname`; while true; do sleep 3600; done
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  volumes:
+  - name: data
+    persistentVolumeClaim:
+      claimName: test
+```
+
+Creating this will give you a pod pinned to the node the scheduler initally picks, with a directory mounted at /data backed by `/var/lib/kubelet/volumes/pv-uuid`. Creating the same thing without the PV section will invoke a dynamic storage provisioner to create an identical PV.
+
+## Alternatives
+
+I (@bprashanth) belive that a sticky emptydir volume as proposed strikes a good balance between hostPath local storage and a spectrum of other possibilities such as:
+* Hostpath dynamic provisioner: The petset team is still considering a short term solution that involves a dynamic provisioner creating a PV backed by hostPath, pointing to the same `/var/lib/kubelet/volumes/pv-uuid` location. This requires 0 code modifications in the node volume subsystem, but requires the scheduler and kubelet admission time changes (relatively simple). This also means the sticky volumes will not get cleaned up, so it is not a viable long term plan.
+* Streamline distributed filesystem deployment on Kubernetes: If we handwave the performance benefits of using local storage (writing to local disk is ~ going over a LAN), we could in theory make it really easy to deploy eg: Gluster and leverage local storage. While this is still a possibility, it introduces another failure mode, and distributed filesystems are far from being easy to deploy. It also significantly raises the bar for the dev/test use case.
+* Use an NFS RC: This is already possible, but still doesn't give us the performance benefit, and is tedious to setup (though not as tedious as a distributed filesystem).
```

Comment:
We call this the dynamic NFS self-provisioner - a dynamic provisioner running on a node (in a pod) that observes the capacity on the node it has been allocated, subdivides that by performing dynamic provisioning (if it observes a RWM PVC created, it eagerly tries to bind to it).  Each provisioner can fight to claim PVCs, and use the API as the synchronization point.

It's been a long running goal to try to get something like that working, and I'm hopeful someone starts it up soon.